### PR TITLE
Error instead of panic in check bind

### LINKED ISSUE #6011: regression: Panic instead of validation error when checking binding compatibility
While updating servo from f25e07b984ab391628d9568296d5970981d79d8b to c0e7c1ef945a7dd61c81fb951ea554213811aee0 in https://github.com/servo/servo/pull/32827, I noticed new crash (also reproducible on latest firefox nightly) on: [webgpu:api,validation,encoding,programmable,pipeline_bind_group_compat:bgl_binding_mismatch:*](https://gpuweb.github.io/cts/standalone/?runnow=1&q=webgpu:api,validation,encoding,programmable,pipeline_bind_group_compat:bgl_binding_mismatch:*) (src of the test: https://github.com/gpuweb/cts/blob/50b6e7a7435e8d1a973cbf67347938ce05188df0/src/webgpu/api/validation/encoding/programmable/pipeline_bind_group_compat.spec.ts#L636)

Description
On f25e07b984ab391628d9568296d5970981d79d8b there is validation error:

```
Validation {
    source: ContextError {
        string: "ComputePass::end",
        cause: ComputePassError {
            scope: Dispatch {
                indirect: false,
            },
            inner: Dispatch(
                IncompatibleBindGroup(
                    IncompatibleBindGroupError {
                        index: 0,
                        pipeline: ResourceErrorIdent {
                            type: "ComputePipeline",
                            label: "",
                        },
                        diff: [
                            "Should be compatible an with an explicit BindGroupLayout with '' label",
                            "Assigned explicit BindGroupLayout with '' label",
                            "Entry 3 not found in assigned bind group layout",
                            "Entry 2 not found in expected bind group layout",
                        ],
                    },
                ),
            ),
        },
        label_key: "label",
        label: "",
    },
    description: "Validation Error\n\nCaused by:\n    In ComputePass::end\n    In a dispatch command, indirect:false\n    In a dispatch command, indirect:false\n    Bind group at index 0 is incompatible with the current set ComputePipeline with '' label\n",
}
```

but c0e7c1ef945a7dd61c81fb951ea554213811aee0 (also on current trunk: 205f1e3ab60a4c8c6b6f901803eb9cfc3a5b62f3) panics on https://github.com/gfx-rs/wgpu/blob/205f1e3ab60a4c8c6b6f901803eb9cfc3a5b62f3/wgpu-core/src/command/bind.rs#L187:
```
Hello, world!
thread 'main' panicked at /home/user/.cargo/git/checkouts/wgpu-53e70f8674b08dd4/c0e7c1e/wgpu-core/src/command/bind.rs:187:70:
called `Option::unwrap()` on a `None` value
stack backtrace:
   0: rust_begin_unwind
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/std/src/panicking.rs:645:5
   1: core::panicking::panic_fmt
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/panicking.rs:72:14
   2: core::panicking::panic
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/panicking.rs:145:5
   3: core::option::unwrap_failed
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/option.rs:1985:5
   4: core::option::Option<T>::unwrap
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/option.rs:933:21
   5: wgpu_core::command::bind::compat::Entry<A>::check
             at /home/user/.cargo/git/checkouts/wgpu-53e70f8674b08dd4/c0e7c1e/wgpu-core/src/command/bind.rs:187:36
   6: wgpu_core::command::bind::compat::BoundBindGroupLayouts<A>::get_invalid
             at /home/user/.cargo/git/checkouts/wgpu-53e70f8674b08dd4/c0e7c1e/wgpu-core/src/command/bind.rs:258:17
   7: wgpu_core::command::bind::Binder<A>::check_compatibility
             at /home/user/.cargo/git/checkouts/wgpu-53e70f8674b08dd4/c0e7c1e/wgpu-core/src/command/bind.rs:432:9
   8: wgpu_core::command::compute::State<A>::is_ready
             at /home/user/.cargo/git/checkouts/wgpu-53e70f8674b08dd4/c0e7c1e/wgpu-core/src/command/compute.rs:238:13
   9: wgpu_core::command::compute::dispatch
             at /home/user/.cargo/git/checkouts/wgpu-53e70f8674b08dd4/c0e7c1e/wgpu-core/src/command/compute.rs:847:5
  10: wgpu_core::command::compute::<impl wgpu_core::global::Global>::compute_pass_end_impl
             at /home/user/.cargo/git/checkouts/wgpu-53e70f8674b08dd4/c0e7c1e/wgpu-core/src/command/compute.rs:586:21
  11: wgpu_core::command::compute::<impl wgpu_core::global::Global>::compute_pass_end
             at /home/user/.cargo/git/checkouts/wgpu-53e70f8674b08dd4/c0e7c1e/wgpu-core/src/command/compute.rs:368:9
  12: <wgpu_core::command::compute::ComputePass<A> as wgpu_core::command::dyn_compute_pass::DynComputePass>::end
             at /home/user/.cargo/git/checkouts/wgpu-53e70f8674b08dd4/c0e7c1e/wgpu-core/src/command/dyn_compute_pass.rs:172:9
  13: <wgpu::backend::wgpu_core::ContextWgpuCore as wgpu::context::Context>::compute_pass_end
             at /home/user/.cargo/git/checkouts/wgpu-53e70f8674b08dd4/c0e7c1e/wgpu/src/backend/wgpu_core.rs:2587:29
  14: <T as wgpu::context::DynContext>::compute_pass_end
             at /home/user/.cargo/git/checkouts/wgpu-53e70f8674b08dd4/c0e7c1e/wgpu/src/context.rs:3306:9
  15: <wgpu::api::compute_pass::ComputePassInner as core::ops::drop::Drop>::drop
             at /home/user/.cargo/git/checkouts/wgpu-53e70f8674b08dd4/c0e7c1e/wgpu/src/api/compute_pass.rs:215:13
  16: core::ptr::drop_in_place<wgpu::api::compute_pass::ComputePassInner>
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/ptr/mod.rs:515:1
  17: core::ptr::drop_in_place<wgpu::api::compute_pass::ComputePass>
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/ptr/mod.rs:515:1
  18: wgpu_problem::run::{{closure}}
             at ./src/main.rs:190:5
  19: pollster::block_on
             at /home/user/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pollster-0.3.0/src/lib.rs:128:15
  20: wgpu_problem::main
             at ./src/main.rs:13:5
  21: core::ops::function::FnOnce::call_once
             at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/ops/function.rs:250:5
```

Repro steps
I was able to reproduce what CTS does by tracing wgpu functions that are called by servo; repro is available here: https://github.com/sagudev/wgpu-problem/tree/empty-bind-error

### PR DESCRIPTION
**Pull request by sagudev**
**Connections**
Fixes #6011

**Description**
Removed zipping of binding entries introduced in https://github.com/gfx-rs/wgpu/commit/4a19ac279c4f81aacedb1d215c884c10fe115275 (to make sure binding numbers actually match) and add unknown error for fallback.

**Testing**
Fixes repro and servo CTS run does not crash anymore.

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `cargo clippy`. If applicable, add:
  - [ ] `--target wasm32-unknown-unknown`
  - [ ] `--target wasm32-unknown-emscripten`
- [x] Run `cargo xtask test` to run tests.
- [ ] Add change to `CHANGELOG.md`. See simple instructions inside file.

### CODE REVIEW

File: wgpu-core/src/command/bind.rs [hunk 1]
```diff
@@ -142,14 +144,21 @@ mod compat {
...
 
-                        for ((&binding, expected_entry), (_, assigned_entry)) in zipped {
+                        for (expected_entry, assigned_entry) in zipped {
+                            if assigned_entry.binding != expected_entry.binding {
+                                errors.push(EntryError::Binding {
+                                    expected: expected_entry.binding,
+                                    assigned: assigned_entry.binding,
+                                });
+                            }
+                            let binding = expected_entry.binding;
```
Comment (sagudev) [inline wgpu-core/src/command/bind.rs]: @teoxoy this is somehow weird, as each entry has different binding number. I think `if assigned_entry.binding != expected_entry.binding {` we should just continue and not print other errors, what do you think?

Comment (teoxoy) [reply wgpu-core/src/command/bind.rs]: Yeah, we shouldn't zip those at all actually.

Comment (teoxoy) [reply wgpu-core/src/command/bind.rs]:
```rust
for expected_entry in expected_bgl.entries.values() {
    if let Some(assigned_entry) =
        assigned_bgl.entries.get(expected_entry.binding)
    {
        if assigned_entry.visibility != expected_entry.visibility {
            errors.push(EntryError::Visibility {
                binding: expected_entry.binding,
                expected: expected_entry.visibility,
                assigned: assigned_entry.visibility,
            });
        }
        if assigned_entry.ty != expected_entry.ty {
            errors.push(EntryError::Type {
                binding: expected_entry.binding,
                expected: expected_entry.ty,
                assigned: assigned_entry.ty,
            });
        }
        if assigned_entry.count != expected_entry.count {
            errors.push(EntryError::Count {
                binding: expected_entry.binding,
                expected: expected_entry.count,
                assigned: assigned_entry.count,
            });
        }
    } else {
        errors.push(EntryError::ExtraExpected {
            binding: expected_entry.binding,
        });
    }
}

for assigned_entry in assigned_bgl.entries.values() {
    if !expected_bgl.entries.contains_key(assigned_entry.binding) {
        errors.push(EntryError::ExtraAssigned {
            binding: assigned_entry.binding,
        });
    }
}
```

I got this working locally.

Comment (sagudev) [reply wgpu-core/src/command/bind.rs]: > Yeah, we shouldn't zip those at all actually.

I just checked in spec and I think we can, see https://gpuweb.github.io/gpuweb/#bind-group-compatibility (either it must exists in both or it must be missing on both)

Comment (sagudev) [reply wgpu-core/src/command/bind.rs]: > I got this working locally.

So basically https://github.com/gfx-rs/wgpu/commit/4a19ac279c4f81aacedb1d215c884c10fe115275#diff-d89ef4ca07007b9044c6ca29c1632570d5e437de81fcaafde8807ebd894d6e9bL81 but with new error types.

Comment (teoxoy) [reply wgpu-core/src/command/bind.rs]: The issue is that if we zip the iterators we'll end up with errors that are not accurate in situations like:
expected_bgl with bindings: 0, 1, 2, 3, 4
assigned_bgl with bindings: 0, 2, 3, 4
Binding 1 will be compared with 2, 2 with 3 and 3 with 4 and we'll say 4 is missing.

If we don't zip, then we will only say that binding 1 is missing.

Comment (sagudev) [reply wgpu-core/src/command/bind.rs]: Ah, you're right.

**Comment by sagudev** [comment]
there is still one failing test
```
• CRASH [expected OK] /_webgpu/webgpu/cts.https.html?q=webgpu:api,validation,encoding,programmable,pipeline_bind_group_compat:default_bind_group_layouts_never_match,render_pass:*
2024-07-22T17:06:01.3143733Z 
2024-07-22T17:06:01.3143816Z       
2024-07-22T17:06:01.3154960Z       called Option::unwrap() on a None value (thread WGPU, at /home/runner/.cargo/git/checkouts/wgpu-4f4c57509d80a789/8a89e85/wgpu-core/src/command/bind.rs:182)
2024-07-22T17:06:01.3155117Z          0: servoshell::backtrace::print
2024-07-22T17:06:01.3155256Z          1: servoshell::panic_hook::panic_hook
2024-07-22T17:06:01.3155391Z          2: std::panicking::rust_panic_with_hook
2024-07-22T17:06:01.3155561Z          3: std::panicking::begin_panic_handler::{{closure}}
2024-07-22T17:06:01.3155749Z          4: std::sys_common::backtrace::__rust_end_short_backtrace
2024-07-22T17:06:01.3155845Z          5: rust_begin_unwind
2024-07-22T17:06:01.3155964Z          6: core::panicking::panic_fmt
2024-07-22T17:06:01.3156072Z          7: core::panicking::panic
2024-07-22T17:06:01.3156178Z          8: core::option::unwrap_failed
2024-07-22T17:06:01.3156458Z          9: wgpu_core::command::bind::compat::BoundBindGroupLayouts<A>::get_invalid
2024-07-22T17:06:01.3156620Z         10: wgpu_core::command::render::State<A>::is_ready
2024-07-22T17:06:01.3156747Z         11: wgpu_core::command::render::draw
2024-07-22T17:06:01.3157126Z         12: <wgpu_core::command::render::RenderPass<A> as wgpu_core::command::dyn_render_pass::DynRenderPass>::end
2024-07-22T17:06:01.3157247Z         13: webgpu::wgpu_thread::WGPU::run
2024-07-22T17:06:01.3157443Z         14: std::sys_common::backtrace::__rust_begin_short_backtrace
2024-07-22T17:06:01.3157622Z         15: core::ops::function::FnOnce::call_once{{vtable.shim}}
2024-07-22T17:06:01.3158008Z         16: std::sys::pal::unix::thread::Thread::new::thread_start
2024-07-22T17:06:01.3158108Z         17: <unknown>
2024-07-22T17:06:01.3158190Z         18: <unknown>
```

**Comment by sagudev** [comment]
Why exactly we do `Ok` only when `is_equal` (which uses Arc::ptr_eq, so it must be same object) instead of doing content comparison (we already do it, but only for error printing => no error should be ok too I think).

**Comment by teoxoy** [comment]
We use `is_equal` in `is_valid` as well, this is an invariant currently because we deduplicate BGLs at creation.

https://github.com/gfx-rs/wgpu/blob/205f1e3ab60a4c8c6b6f901803eb9cfc3a5b62f3/wgpu-core/src/device/global.rs#L800-L810

**Comment by teoxoy** [comment]
I looked at the `default_bind_group_layouts_never_match` test and it's not clear to me what we are missing. Is the `.unwrap()` failing on all the subtests?

**Comment by sagudev** [comment]
> I looked at the `default_bind_group_layouts_never_match` test and it's not clear to me what we are missing. Is the `.unwrap()` failing on all the subtests?

No, some works, I am currently writing repro in rust using wgpu (also causes crash in firefox: https://gpuweb.github.io/cts/standalone/?q=webgpu:api,validation,encoding,programmable,pipeline_bind_group_compat:default_bind_group_layouts_never_match,render_pass:pipelineType=%22auto0%22;bindingType=%22auto0%22;swap=true;empty=false;renderCommand=%22draw%22) I think invariant might be broken when using `get_bind_group_layout`, because I debug printed both bgls and they have same content except for ptrs (raw field and tracker index), diff:
```patch
@@ -1,7 +1,7 @@
- expected_bgl BindGroupLayout {
+ actual_bgl BindGroupLayout {
   │     raw: Some(
   │         BindGroupLayout {
-  │             raw: 0x7f07d052a780,
+  │             raw: 0x7f07d052a830,
   │             desc_count: DescriptorTotalCount {
   │                 sampler: 0,
   │                 combined_image_sampler: 0,
@@ -133,7 +133,7 @@ expected_bgl BindGroupLayout {
   │     label: "",
   │     tracking_data: TrackingData {
   │         tracker_index: TrackerIndex(
-  │             68,
+  │             69,
   │         ),
   │         tracker_indices: SharedTrackerIndexAllocator {
   │             inner: Mutex {
```

**Comment by teoxoy** [comment]
That can happen if the BGLs differ in `exclusive_pipeline`, if they don't then it's a bug somewhere.

**Comment by sagudev** [comment]
> That can happen if the BGLs differ in `exclusive_pipeline`

This should already be catched above, but it's worth rechecking. EDIT: It's okay.

**Comment by sagudev** [comment]
repro: https://github.com/sagudev/wgpu-problem/tree/another-empty-bind-panic-2 (this took way to much time, because https://github.com/sagudev/wgpu-problem/commit/2da4ad294d3c481f24a49c7cb94eec862de96eb2)

**Comment by sagudev** [comment]
Same commit is bad one, expected error instead of crash:
```
ERROR: Validation {
    source: ContextError {
        string: "RenderPass::end",
        cause: RenderPassError {
            scope: Draw {
                kind: Draw,
                indexed: false,
            },
            inner: Draw(
                IncompatibleBindGroup(
                    IncompatibleBindGroupError {
                        index: 2,
                        pipeline: ResourceErrorIdent {
                            type: "RenderPipeline",
                            label: "",
                        },
                        diff: [
                            "Should be compatible an with an implicit BindGroupLayout with '' label",
                            "Assigned implicit BindGroupLayout with '' label",
                        ],
                    },
                ),
            ),
        },
        label_key: "label",
        label: "",
    },
    description: "Validation Error\n\nCaused by:\n    In RenderPass::end\n    In a draw command, kind: Draw\n    In a draw command, kind: Draw\n    Bind group at index 2 is incompatible with the current set RenderPipeline with '' label\n",
}
```

**Comment by sagudev** [comment]
Before https://github.com/gfx-rs/wgpu/commit/4a19ac279c4f81aacedb1d215c884c10fe115275 there was always something in the diff (even if we didn't know exactly what didn't match), so I think adding fallback using `unwrap_or` would me the most appropriate solution (for now, as the root cause is for separate issue).

**Comment by sagudev** [comment]
rebased and added changelog entry

**Comment by teoxoy** [comment]
> repro: https://github.com/sagudev/wgpu-problem/tree/another-empty-bind-panic-2 (this took way to much time, because https://github.com/sagudev/wgpu-problem/commit/2da4ad294d3c481f24a49c7cb94eec862de96eb2)

Thanks for looking into it and putting up a minimal reproducible example!

I'll look at it and try to come up with a fix this week. Part of the reason I added the `.unwrap()` is to catch cases like this where something we didn't expect happened. This would have gone unnoticed for a lot longer otherwise.

Let's land this as is for now to get it in the patch release.

**Review by teoxoy — APPROVED** [review]

**Comment by sagudev** [comment]
> > repro: https://github.com/sagudev/wgpu-problem/tree/another-empty-bind-panic-2 (this took way to much time, because [sagudev/wgpu-problem@2da4ad2](https://github.com/sagudev/wgpu-problem/commit/2da4ad294d3c481f24a49c7cb94eec862de96eb2))
> 
> Thanks for looking into it and putting up a minimal reproducible example!
> 
> I'll look at it and try to come up with a fix this week. Part of the reason I added the `.unwrap()` is to catch cases like this where something we didn't expect happened. This would have gone unnoticed for a lot longer otherwise.

I will open new issue for this, with all known info in one place.

**Pull request merged**